### ISSUE TITLE
RBAC: access denied on minio artifacts in KFP

### ISSUE THREAD

**Issue by pbasov** [deployKF/deployKF#98]
### Checks

- [X] I have searched the [existing issues](https://github.com/deployKF/deployKF).
- [X] This issue is NOT specific to the CLI. (If so, please open an issue on the [CLI repo](https://github.com/deployKF/cli/issues))

### deployKF Version

0.1.3

### Kubernetes Version

```shell
Client Version: v1.29.0
Server Version: v1.29.0
```

### Description

KFP artifacts stored in Minio are not available from KFP UI, throwing `RBAC: access denied` on trying to access `/pipeline/artifacts/minio/kubeflow-pipelines/artifacts/...`.

They are, however accessible with the same user from Minio console.
I'm using plain passwords to log in for now.

**Comment by thesuperzapper** [comment]
@pbasov can you please confirm if the step you are getting an RBAC issue was "cached" because it was first run in another namespace/profile?

deployKF will try and redirect you to the correct namespace, but if your logged in user does not have permission on the correct namespace, you will get an RBAC error.

**Comment by pbasov** [comment]
I'm a single user (apart from admin@example.com) in a single profile. I ran the pipeline from KF UI.
```
deploykf_auth:
  dex:
    staticPasswords:
    - email: "admin@example.com"
      password:
        value: ""
    - email: "mail@weystrom.net"
      password:
        value: ""
```
```
deploykf_profiles_generator:
  users:
    - id: weystrom
      email: "mail@weystrom.net"
  groups:
    - id: ml-team
      users:
        - weystrom
  profiles:
    - name: mirantis-ml-profile
      members:
        - group: ml-team
          access:
            role: edit
            notebooksAccess: true
```

**Comment by pbasov** [comment]
```[2024-02-06T17:14:43.356Z] "GET /pipeline/artifacts/minio/kubeflow-pipelines/artifacts/ml-profile/hello-world-pipeline-b7s4n/2024/02/06/hello-world-pipeline-b7s4n-2915942301/main.log HTTP/1.1" 403 - rbac_access_denied_matched_policy[none] - "-" 0 19 0 - "192.168.80.3" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:122.0) Gecko/20100101 Firefox/122.0" "1f1dd94e-a426-4485-96a9-f43460ded4a6" "ml-pipeline-ui-artifact.ml-profile:80" "-" inbound|3000|| - 10.0.2.251:3000 192.168.80.3:0 - -```

Am I missing some generated authpolicy from `kyverno` perhaps?

**Comment by thesuperzapper** [comment]
@pbasov I assume that you have synced all apps fully in ArgoCD, and either way, try re-syncing them (with the sync script).

Also, please list out all the `AuthorizationPolicies` in your `ml-profile` namespace (YAML preferably).

**Comment by pbasov** [comment]
Just re-synced everything, no change.

<details>
  <summary>AuthorizationPolicies </summary>

```
apiVersion: v1
items:
- apiVersion: security.istio.io/v1
  kind: AuthorizationPolicy
  metadata:
    annotations:
      argocd.argoproj.io/compare-options: IgnoreExtraneous
      argocd.argoproj.io/sync-options: Prune=false
      argocd.argoproj.io/tracking-id: kf-tools--pipelines:security.istio.io/AuthorizationPolicy:ml-profile/ml-pipeline-visualizationserver
    creationTimestamp: "2024-02-04T22:50:34Z"
    generation: 1
    labels:
      app.kubernetes.io/instance: kf-tools--pipelines
      app.kubernetes.io/managed-by: kyverno
      generate.kyverno.io/policy-name: kubeflow-pipelines--generate-profile-resources
      generate.kyverno.io/policy-namespace: ""
      generate.kyverno.io/rule-name: AuthorizationPolicy--ml-pipeline-visualizationserver
      generate.kyverno.io/trigger-apiversion: v1
      generate.kyverno.io/trigger-kind: Namespace
      generate.kyverno.io/trigger-name: ml-profile
      generate.kyverno.io/trigger-namespace: ""
    name: ml-pipeline-visualizationserver
    namespace: ml-profile
    resourceVersion: "665612"
    uid: 6cad128c-4094-4256-80a7-b80b7dbcd9dd
  spec:
    rules:
    - from:
      - source:
          principals:
          - cluster.local/ns/kubeflow/sa/ml-pipeline
    selector:
      matchLabels:
        app: ml-pipeline-visualizationserver
- apiVersion: security.istio.io/v1
  kind: AuthorizationPolicy
  metadata:
    annotations:
      role: admin
      user: admin@example.com
    creationTimestamp: "2024-02-04T21:49:54Z"
    generation: 1
    name: ns-owner-access-istio
    namespace: ml-profile
    ownerReferences:
    - apiVersion: kubeflow.org/v1
      blockOwnerDeletion: true
      controller: true
      kind: Profile
      name: ml-profile
      uid: 4a0b367e-185c-4991-9d3c-76a1a575e9d6
    resourceVersion: "637996"
    uid: 3efa6c72-d7a4-462a-851f-c78f54c3cc46
  spec:
    rules:
    - when:
      - key: request.headers[kubeflow-userid]
        values:
        - admin@example.com
    - when:
      - key: source.namespace
        values:
        - ml-profile
    - to:
      - operation:
          paths:
          - /healthz
          - /metrics
          - /wait-for-drain
    - from:
      - source:
          principals:
          - cluster.local/ns/kubeflow/sa/notebook-controller-service-account
      to:
      - operation:
          methods:
          - GET
          paths:
          - '*/api/kernels'
- apiVersion: security.istio.io/v1
  kind: AuthorizationPolicy
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"security.istio.io/v1beta1","kind":"AuthorizationPolicy","metadata":{"annotations":{},"labels":{"app.kubernetes.io/instance":"dkf-core--deploykf-profiles-generator","app.kubernetes.io/managed-by":"Helm","app.kubernetes.io/name":"deploykf-profiles-generator","helm.sh/chart":"deploykf-profiles-generator-1.0.0"},"name":"ns-owner-access-istio--override","namespace":"ml-profile"},"spec":{"action":"DENY","rules":[{"from":[{"source":{"notPrincipals":["cluster.local/ns/deploykf-istio-gateway/sa/deploykf-gateway","cluster.local/ns/kubeflow/sa/ml-pipeline-ui"]}}],"when":[{"key":"request.headers[kubeflow-userid]","values":["admin@example.com"]}]}]}}
    creationTimestamp: "2024-02-04T21:49:57Z"
    generation: 3
    labels:
      app.kubernetes.io/instance: dkf-core--deploykf-profiles-generator
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: deploykf-profiles-generator
      helm.sh/chart: deploykf-profiles-generator-1.0.0
    name: ns-owner-access-istio--override
    namespace: ml-profile
    resourceVersion: "1643490"
    uid: 4317dc7a-1af4-41f3-bf0c-fda59971363e
  spec:
    action: DENY
    rules:
    - from:
      - source:
          notPrincipals:
          - cluster.local/ns/deploykf-istio-gateway/sa/deploykf-gateway
          - cluster.local/ns/kubeflow/sa/ml-pipeline-ui
      when:
      - key: request.headers[kubeflow-userid]
        values:
        - admin@example.com
- apiVersion: security.istio.io/v1
  kind: AuthorizationPolicy
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"security.istio.io/v1beta1","kind":"AuthorizationPolicy","metadata":{"annotations":{"role":"edit","user":"mail@example.net"},"labels":{"app.kubernetes.io/instance":"dkf-core--deploykf-profiles-generator","app.kubernetes.io/managed-by":"Helm","app.kubernetes.io/name":"deploykf-profiles-generator","helm.sh/chart":"deploykf-profiles-generator-1.0.0"},"name":"user-mail-weystrom-net-clusterrole-edit","namespace":"ml-profile"},"spec":{"rules":[{"from":[{"source":{"principals":["cluster.local/ns/deploykf-istio-gateway/sa/deploykf-gateway","cluster.local/ns/kubeflow/sa/ml-pipeline-ui"]}}],"when":[{"key":"request.headers[kubeflow-userid]","values":["mail@example.net"]}]}]}}
      role: edit
      user: mail@example.net
    creationTimestamp: "2024-02-04T21:49:57Z"
    generation: 1
    labels:
      app.kubernetes.io/instance: dkf-core--deploykf-profiles-generator
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: deploykf-profiles-generator
      helm.sh/chart: deploykf-profiles-generator-1.0.0
    name: user-mail-example-net-clusterrole-edit
    namespace: ml-profile
    resourceVersion: "638084"
    uid: a996c577-11a5-422f-994d-f4f96496d1cf
  spec:
    rules:
    - from:
      - source:
          principals:
          - cluster.local/ns/deploykf-istio-gateway/sa/deploykf-gateway
          - cluster.local/ns/kubeflow/sa/ml-pipeline-ui
      when:
      - key: request.headers[kubeflow-userid]
        values:
        - mail@example.net
kind: List
metadata:
  resourceVersion: ""
```
</details>

**Comment by thesuperzapper** [comment]
@pbasov it looks like somehow you have ended up with `mail@example.net` rather than `weystrom@example.net` as the email for your other user.

Check your values file (or files if there are more than one), and check that `deploykf_core.deploykf_profiles_generator.users` does not have any unexpected values.

**Comment by pbasov** [comment]
> @pbasov it looks like somehow you have ended up with mail@example.net rather than weystrom@example.net as the email for your other user.

Ah, that's just me poorly sanitizing real values.

I'm wondering if I shot myself in the foot deploying k8s with a different cluster domain instead of `cluster.local` and if that affects `trustDomain` of Istio, even though everything else is working. I changed CoreDNS config to include `cluster.local` so deployKF would work (since cluster.local is pretty much hardcoded here) and everything else seems to be working.
https://istio.io/latest/docs/tasks/security/authorization/authz-td-migration/

I'll fix cluster.local in the cluster and check again. Will close for now and report back if that wasn't it.

**Comment by thesuperzapper** [comment]
@pbasov you are correct, deployKF currently requires the cluster domain to be `cluster.local`, so that was likely the problem.

**Comment by pbasov** [comment]
@thesuperzapper That wasn't it, i've redeployed the cluster with `cluster.local` clusterDomain and still having the same issue. Which authPolicy is supposed to be working here?

**Comment by thesuperzapper** [comment]
@pbasov The `user-mail-example-net-clusterrole-edit` (or whatever its actually called) is the one which should give you access, but it's clearly not correctly matching given your `rbac_access_denied_matched_policy[none]` error.

Can you confirm if you are trying to expose anything else on the same istio gateway as deployKF?

Or if you have any other Istio stuff/configs on this cluster?

**Comment by pbasov** [comment]
So I've switched network underlay of all things, and that fixed it. I'm a bit puzzled why to be honest, but now everything works. My guess is cross-node communication was broken somehow, but that was the only symptom. I'll try to reproduce it later out of curiousity, but it's definitely not a provisioning issue.

Thanks for all the help, and my compliments on this Argo setup, takes less than an hour to bring KF up from scratch with deployKF.

**Comment by thesuperzapper** [comment]
@pbasov thanks for the feedback! 

Yes, we are planning to include more tools, and KServe is top of that list right now!

**Comment by metalwhale** [comment]
First, I really appreciate your work on developing deployKF. Thank you so much for doing this!

I came across this issue while searching for a similar problem, and I want to share my experience with others who may encounter it. In my case, I got the following error message:
```
Failed to get object in bucket kubeflow-pipelines at path artifacts/xxx/.../xxx/mlpipeline-ui-metadata.tgz: S3Error: The Access Key Id you provided does not exist in our records.
```
It turns out that I had enabled minio **after** creating the profile, so it seems like `ml-pipeline-ui-artifact` didn't update the correct permission. After manually deleting the pod and resyncing, everything works like a charm (:

**Comment by thesuperzapper** [comment]
@metalwhale for reference, with the new approach to generating those resources in 0.1.4, re-syncing the apps would probably have been sufficient.

We no longer use Kyverno to generate these profile KFP responses, and instead just include the resources for each profile in the ArgoCD app directly.

**Comment by harsha1163** [comment]
@thesuperzapper, we encountered the same issue which @metalwhale has faced earlier. His fix helped us. We already did an argo sync using the shell script that was provided after redeploying the minio. but still we faced this issue. 

May be a force restart of ml-pipeline-ui-artifact pods would help while argo sync happens.

**Comment by thesuperzapper** [comment]
@harsha1163 I have a suspicion that this was caused by a known error where the "automatic restart on secret change" does not work for very long secret names, like those that occur in each profile for Kubeflow Pipelines called `cloned--xxx` (https://github.com/kyverno/kyverno/issues/9134).

This should be resolved in deployKF `0.1.6`, where we plan to update Kyverno to a version which does not have this problem. However, until that time, you might need to manually restart those `ml-pipeline-...` deployments if you change your object store secrets.
